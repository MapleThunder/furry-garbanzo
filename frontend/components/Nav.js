import Link from "next/link";
import NavStyles from "./styles/NavStyles";

export const Nav = () => (
  <NavStyles>
    <Link href="/">
      <a>Items</a>
    </Link>
    <Link href="/sell">
      <a>Sell</a>
    </Link>
    <Link href="/signup">
      <a>Sign Up</a>
    </Link>
    <Link href="/orders">
      <a>orders</a>
    </Link>
    <Link href="/me">
      <a>Account</a>
    </Link>
  </NavStyles>
);
