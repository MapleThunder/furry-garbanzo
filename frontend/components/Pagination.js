import React from "react";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import Head from "next/head";
import Link from "next/link";
import PaginationStyles from "./styles/PaginationStyles";
import { perPage } from "../config";

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemCount
  }
`;

export const Pagination = ({ page }) => {
  return (
    <Query query={PAGINATION_QUERY}>
      {({ data, loading, error }) => {
        if (loading) return <p>Loading...</p>;
        const count = data.itemCount;
        const pages = Math.ceil(count / perPage);

        return (
          <PaginationStyles data-testid="pagination">
            <Head>
              <title>
                Sick Fits â€” Page {page} of {pages}
              </title>
            </Head>
            <Link
              href={{
                pathname: "items",
                query: { page: page - 1 },
              }}
            >
              <a className="prev" aria-disabled={page <= 1}>
                &larr; Prev
              </a>
            </Link>
            <p>
              Page {page} of <span className="totalPages">{pages}</span>
            </p>
            <p>{count} Items Total</p>
            <Link
              href={{
                pathname: "items",
                query: { page: page + 1 },
              }}
            >
              <a className="next" aria-disabled={page >= pages}>
                Next &rarr;
              </a>
            </Link>
          </PaginationStyles>
        );
      }}
    </Query>
  );
};
