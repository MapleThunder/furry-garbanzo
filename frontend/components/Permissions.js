import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";

import Error from "./ErrorMessage";
import Table from "./styles/Table";
import SickButton from "./styles/SickButton";

const possiblePermissions = [
  "ADMIN",
  "USER",
  "ITEM_CREATE",
  "ITEM_UPDATE",
  "ITEM_DELETE",
  "PERMISSION_UPDATE",
];

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      id
      name
      email
      permissions
    }
  }
`;

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation UPDATE_PERMISSIONS_MUTATION(
    $permissions: [Permission]
    $userId: ID!
  ) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      permissions
      name
      email
    }
  }
`;

export const Permissions = (props) => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => (
      <div>
        <Error error={error} />
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map((permission) => (
                  <th key={permission}>{permission}</th>
                ))}
                <th>👇🏻</th>
              </tr>
            </thead>
            <tbody>
              {data.users.map((user) => (
                <UserRow key={user.id} user={user} />
              ))}
            </tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
);

class UserRow extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.number,
      permissions: PropTypes.array,
    }).isRequired,
  };
  state = {
    permissions: this.props.user.permissions,
  };

  handlePermissionChange = (e) => {
    const checkbox = e.target;
    // take a copy of current permissions
    let updatedPermissions = [...this.state.permissions];
    // Figure out whether permissions need to be added or removed
    if (checkbox.checked) {
      // Add permission
      updatedPermissions.push(checkbox.value);
    } else {
      updatedPermissions = updatedPermissions.filter(
        (permission) => permission !== checkbox.value
      );
    }
    this.setState({ permissions: updatedPermissions });
  };

  render() {
    const { user } = this.props;
    return (
      <Mutation
        mutation={UPDATE_PERMISSIONS_MUTATION}
        variables={{
          permissions: this.state.permissions,
          userId: this.props.user.id,
        }}
      >
        {(updatePermissions, { loading, error }) => (
          <>
            {error && (
              <tr>
                <td colSpan="8">
                  <Error error={error} />
                </td>
              </tr>
            )}
            <tr>
              <td>{user.name}</td>
              <td>{user.email}</td>
              {possiblePermissions.map((permission) => (
                <td key={`${user.id}-permission-${permission}`}>
                  <label htmlFor={`${user.id}-permission-${permission}`}>
                    <input
                      id={`${user.id}-permission-${permission}`}
                      type="checkbox"
                      checked={this.state.permissions.includes(permission)}
                      value={permission}
                      onChange={this.handlePermissionChange}
                    />
                  </label>
                </td>
              ))}
              <td>
                <SickButton
                  type="button"
                  disabled={loading}
                  aria-busy={loading}
                  onClick={updatePermissions}
                >
                  Updat{loading ? "ing" : "e"}
                </SickButton>
              </td>
            </tr>
          </>
        )}
      </Mutation>
    );
  }
}
