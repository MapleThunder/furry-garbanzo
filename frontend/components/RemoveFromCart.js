import React from "react";
import { Mutation } from "react-apollo";
import styled from "styled-components";
import PropTypes from "prop-types";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./User";

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${(props) => props.theme.red};
    cursor: pointer;
  }
`;

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

export const RemoveFromCart = ({ id }) => {
  // This is called when a response is returned from the server
  // after performing the mutation
  const update = (cache, payload) => {
    // 1. Read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    // 2. Remove item from cart
    const cartItemId = payload.data.removeFromCart.id;
    data.me.cart = data.me.cart.filter(
      (cartItem) => cartItem.id !== cartItemId
    );
    // 3. Write back to cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  };

  return (
    <Mutation
      mutation={REMOVE_FROM_CART_MUTATION}
      variables={{ id }}
      update={update}
      optimisticResponse={{
        __typename: "Mutation",
        removeFromCart: {
          __typename: "CartItem",
          id,
        },
      }}
    >
      {(removeFromCart, { loading, error }) => {
        return (
          <BigButton
            title="Delete Item"
            disabled={loading}
            onClick={() => {
              removeFromCart().catch((err) => alert(err.message));
            }}
          >
            &times;
          </BigButton>
        );
      }}
    </Mutation>
  );
};

RemoveFromCart.propTypes = {
  id: PropTypes.number.isRequired,
};
