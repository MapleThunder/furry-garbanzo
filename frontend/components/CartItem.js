import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import formatMoney from "../lib/formatMoney";
import { RemoveFromCart } from "./RemoveFromCart";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightGrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`;

export const CartItem = ({ cartItem }) => {
  if (!cartItem.Item) {
    return (
      <CartItemStyles>
        <p>This item has been removed.</p>
        <RemoveFromCart id={cartItem.id} />
      </CartItemStyles>
    );
  }
  return (
    <CartItemStyles>
      <img width="100px" src={cartItem.Item.image} alt={cartItem.Item.title} />
      <div className="cart-item-details">
        <h3>{cartItem.Item.title}</h3>
        <p>
          {formatMoney(cartItem.Item.price * cartItem.quantity)}
          {" - "}
          <em>
            {cartItem.quantity} &times; {formatMoney(cartItem.Item.price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={parseFloat(cartItem.id)} />
    </CartItemStyles>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired,
};
