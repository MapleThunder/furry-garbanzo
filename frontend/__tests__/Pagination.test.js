import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import Router from "next/router";
import { MockedProvider } from "react-apollo/test-utils";
import { Pagination, PAGINATION_QUERY } from "../components/Pagination";

Router.router = {
  push() {},
  prefetch() {},
};

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemCount: length,
        },
      },
    },
  ];
}

describe("<Pagination />", () => {
  it("displays a loading message", () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain("Loading...");
  });

  it("renders paggination for 18 items", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find("span.totalPages").text()).toBe("5");
    const pagination = wrapper.find("[data-testid='pagination']");
    expect(toJSON(pagination)).toMatchSnapshot();
  });

  it("disables prev button on first page", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(true);
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(false);
  });
  it("disables next button on last page", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(false);
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(true);
  });
  it("enables all buttons on middle pages", async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find("a.prev").prop("aria-disabled")).toEqual(false);
    expect(wrapper.find("a.next").prop("aria-disabled")).toEqual(false);
  });
});
