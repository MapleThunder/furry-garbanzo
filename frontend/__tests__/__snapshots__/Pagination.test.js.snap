// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders paggination for 18 items 1`] = `
Array [
  <PaginationStyles
    data-testid="pagination"
  >
    <div
      className="PaginationStyles-aduuar-0 ewJsCc"
      data-testid="pagination"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 0,
            },
          }
        }
      >
        <a
          aria-disabled={true}
          className="prev"
          href="items?page=0"
          onClick={[Function]}
        >
          ← Prev
        </a>
      </Link>
      <p>
        Page 
        1
         of 
        <span
          className="totalPages"
        >
          5
        </span>
      </p>
      <p>
        18
         Items Total
      </p>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 2,
            },
          }
        }
      >
        <a
          aria-disabled={false}
          className="next"
          href="items?page=2"
          onClick={[Function]}
        >
          Next →
        </a>
      </Link>
    </div>
  </PaginationStyles>,
  <div
    className="PaginationStyles-aduuar-0 ewJsCc"
    data-testid="pagination"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 0,
          },
        }
      }
    >
      <a
        aria-disabled={true}
        className="prev"
        href="items?page=0"
        onClick={[Function]}
      >
        ← Prev
      </a>
    </Link>
    <p>
      Page 
      1
       of 
      <span
        className="totalPages"
      >
        5
      </span>
    </p>
    <p>
      18
       Items Total
    </p>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 2,
          },
        }
      }
    >
      <a
        aria-disabled={false}
        className="next"
        href="items?page=2"
        onClick={[Function]}
      >
        Next →
      </a>
    </Link>
  </div>,
]
`;
